File: java/src/main/java/benchmarks/Sieve.java
package benchmarks;

import java.util.Arrays;

public class Sieve extends Benchmark {
    private long limit;
    private long checksum;

    public Sieve() {
        limit = configVal("limit");
        checksum = 0L;
    }

    @Override
    public String name() {
        return "Etc::Sieve";
    }

    @Override
    public void run(int iterationId) {
        int lim = (int) limit;
        byte[] primes = new byte[lim + 1];
        Arrays.fill(primes, (byte) 1);
        primes[0] = 0;
        primes[1] = 0;

        int sqrtLimit = (int) Math.sqrt(lim);

        for (int p = 2; p <= sqrtLimit; p++) {
            if (primes[p] == 1) {
                for (int multiple = p * p; multiple <= lim; multiple += p) {
                    primes[multiple] = 0;
                }
            }
        }

        int lastPrime = 2;
        int count = 1;

        for (int n = 3; n <= lim; n += 2) {
            if (primes[n] == 1) {
                lastPrime = n;
                count++;
            }
        }

        checksum += (lastPrime + count);
    }

    @Override
    public long checksum() {
        return checksum;
    }
}

File: scala/src/main/scala/benchmarks/Sieve.scala
package benchmarks

import java.{util => ju}

class Sieve extends Benchmark:
  private var limit: Long = 0L
  private var checksumVal: Long = 0L

  override def name(): String = "Etc::Sieve"

  override def prepare(): Unit =
    limit = configVal("limit")

  override def run(iterationId: Int): Unit =
    val lim = limit.toInt
    val primes = new Array[Byte](lim + 1)
    ju.Arrays.fill(primes, 1.toByte)
    primes(0) = 0
    primes(1) = 0

    val sqrtLimit = math.sqrt(lim).toInt

    var p = 2
    while p <= sqrtLimit do
      if primes(p) == 1 then
        var multiple = p * p
        while multiple <= lim do
          primes(multiple) = 0
          multiple += p
      p += 1

    var lastPrime = 2
    var count = 1

    var n = 3
    while n <= lim do
      if primes(n) == 1 then
        lastPrime = n
        count += 1
      n += 2

    checksumVal = (checksumVal + lastPrime + count) & 0xffffffffL

  override def checksum(): Long = checksumVal
